wagtail/core/actions/copy_page.py

```diff
@@ -0,0 +1,257 @@
+import json
+import logging
+import uuid
+
+from django.core.exceptions import PermissionDenied
+from modelcluster.models import get_all_child_relations
+
+from wagtail.core.log_actions import log
+from wagtail.core.models.copying import _copy, _copy_m2m_relations
+from wagtail.core.models.i18n import TranslatableMixin
+from wagtail.core.signals import page_published
+
+
+logger = logging.getLogger('wagtail.core')
+
+
+class CopyPageIntegrityError(RuntimeError):
+    """
+    Raised when the page copy cannot be performed for data integrity reasons.
+    """
+    pass
+
+
+class CopyPagePermissionError(PermissionDenied):
+    """
+    Raised when the page copy cannot be performed due to insufficient permissions.
+    """
+    pass
+
+
+class CopyPageAction:
+    """
+    Copies pages and page trees.
+    """
+    def __init__(self, page, to=None, update_attrs=None, exclude_fields=None, recursive=False, copy_revisions=True, keep_live=True, user=None, process_child_object=None, log_action='wagtail.copy', reset_translation_key=True):
+        # Note: These four parameters don't apply to any copied children
+        self.page = page
+        self.to = to
+        self.update_attrs = update_attrs
+        self.exclude_fields = exclude_fields
+
+        self.recursive = recursive
+        self.copy_revisions = copy_revisions
+        self.keep_live = keep_live
+        self.user = user
+        self.process_child_object = process_child_object
+        self.log_action = log_action
+        self.reset_translation_key = reset_translation_key
+
+    def check(self, skip_permission_checks=False):
+        from wagtail.core.models import UserPagePermissionsProxy
+
+        # Essential data model checks
+        if self.page._state.adding:
+            raise CopyPageIntegrityError('Page.copy() called on an unsaved page')
+
+        if self.to and self.recursive and (self.to.id == self.page.id or self.to.is_descendant_of(self.page)):
+            raise CopyPageIntegrityError("You cannot copy a tree branch recursively into itself")
+
+        # Permission checks
+        if self.user and not skip_permission_checks:
+            to = self.to
+            if to is None:
+                to = self.page.get_parent()
+
+            if not self.page.permissions_for_user(self.user).can_copy_to(to, self.recursive):
+                raise CopyPagePermissionError("You do not have permission to copy this page")
+
+            if self.keep_live:
+                destination_perms = UserPagePermissionsProxy(self.user).for_page(self.to)
+
+                if not destination_perms.can_publish_subpage():
+                    raise CopyPagePermissionError("You do not have permission publish a page at the destination")
+
+    def _copy_page(self, page, to=None, update_attrs=None, exclude_fields=None, _mpnode_attrs=None):
```

Comment:
I feel like this or `execute` need wrapping in a transaction - we're doing a lot of successive database operations, we should probably make sure we don't end up with pages without their M2M relations or similar

Reply:
Feel like should be added to ``execute()`` since adding it to ``_copy_page`` would create a save point before copying each page